tower-ready-cache/src/service.rs

```diff
@@ -0,0 +1,380 @@
+use crate::error;
+use futures::{future, stream, Async, Future, Poll, Stream};
+pub use indexmap::Equivalent;
+use indexmap::IndexMap;
+use log::{debug, trace};
+use std::hash::Hash;
+use tokio_sync::oneshot;
+use tower_service::Service;
+use tower_util::Ready;
+
+/// Caches services, segregating ready and unready services.
+#[derive(Debug)]
+pub struct ReadyCache<K, S, Req>
+where
+    K: Eq + Hash,
+{
+    next_ready_index: Option<usize>,
+    ready: IndexMap<K, S>,
+    pending: stream::FuturesUnordered<EvictableReady<K, S, Req>>,
+    cancelations: IndexMap<K, oneshot::Sender<()>>,
+}
+
+#[derive(Debug)]
+enum PendingError<K, E> {
+    Canceled(K),
+    Inner(K, E),
+}
+
+/// A Future that becomes satisfied when an `S`-typed service is ready.
+///
+/// May fail due to cancelation, i.e. if the service is removed from discovery.
+#[derive(Debug)]
+struct EvictableReady<K, S, Req> {
```

Comment:
Isn't eviction specifically for _unready_ services?

Reply:
`EvictableReady` wraps the `tower_util::Ready` future to interact with `evict`.

Reply:
I think there is some overloading of terms here with "ready" also referring to a _service that is ready_. Perhaps just `UnreadyService`?